src/protocol/common/hex.rs

```diff
@@ -168,7 +168,34 @@ pub fn decode_hex_buf(base_buf: &mut [u8]) -> Result<&mut [u8], DecodeHexBufErro
     Ok(&mut base_buf[..decoded_len + odd_adust])
 }
 
-#[allow(dead_code)]
+#[derive(Debug)]
+pub enum DecodeBinBufError {
+    UnexpectedEnd,
+}
+
+/// Decode GDB escaped binary bytes into origin bytes _in place_.
+pub fn decode_bin_buf(buf: &mut [u8]) -> Result<&mut [u8], DecodeBinBufError> {
```

Comment:
I took a bit of time to play with this in godbolt to see if I could optimize away the bounds checks. Here's what I came up with:

https://godbolt.org/z/a65sj96h7

Feel free to riff on it and see if you can simplify it some more.

If that extra "trick the optimizer" block is a bit too sketchy for your tastes, we can swap out the tail end of the function with:

```rust
// by inspection, the value of j will never exceed buf.len().
// unfortunately, the LLVM optimizer isn't smart enough to see this, so
// we have to manually elide the bounds check...
if cfg!(feature = "paranoid_unsafe") {
    Ok(&mut buf[..j])
} else {
    debug_assert(j <= buf.len());
    unsafe { Ok(buf.get_unchecked_mut(..j)) }
}
```